server/routerlicious/packages/services-telemetry/src/lumber.ts

```diff
@@ -0,0 +1,149 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { LumberEventName } from "./lumberEventNames";
+import { LogLevel, LumberType, ITelemetryMetadata, ILumberjackEngine } from "./resources";
+
+// Lumber represents the telemetry data being captured, and it uses a list of
+// ILumberjackEngine to emit the data according to the engine implementation.
+// Lumber should be created through Lumberjack. Additional properties can be added through
+// addProperty(). Once the telemetry event is complete, the user must call either success()
+// or error() on Lumber to emit the data.
+export class Lumber<T extends string = LumberEventName> {
+    private readonly _eventName: T;
+    private readonly _properties: Map<string, any>;
+    private readonly _timestamp: string;
+    private readonly startTime: number;
+    private readonly _type: LumberType;
+    private _metadata: ITelemetryMetadata | undefined;
+    private _latencyInMs: number | undefined;
+    private _successful: boolean | undefined;
+    private _message: string | undefined;
+    private _statusCode: string | undefined;
+    private _exception: Error | undefined;
+    private _logLevel: LogLevel | undefined;
+    private completed: boolean;
+    private readonly engineList: ILumberjackEngine[];
+
+    public get eventName(): T {
+        return this._eventName;
+    }
+
+    public get metadata(): ITelemetryMetadata | undefined {
+        return this._metadata;
+    }
+
+    public get properties(): Map<string, any> {
+        return this._properties;
+    }
+
+    public get type(): LumberType {
+        return this._type;
+    }
+
+    public get timestamp(): string {
+        return this._timestamp;
+    }
+
+    public get latencyInMs(): number | undefined {
+        if (this._type === LumberType.Log) {
+            return undefined;
+        }
+        return this._latencyInMs;
+    }
+
+    public get successful(): boolean | undefined {
+        if (this._type === LumberType.Log) {
+            return undefined;
+        }
+        return this._successful;
+    }
+
+    public get message(): string | undefined {
+        return this._message;
+    }
+
+    public get statusCode(): string | undefined {
+        return this._statusCode;
+    }
+
+    public get exception(): Error | undefined {
+        return this._exception;
+    }
+
+    public get logLevel(): LogLevel | undefined {
+        return this._logLevel;
+    }
+
+    constructor(
+        eventName: T,
+        type: LumberType,
+        engineList: ILumberjackEngine[]) {
+        this._eventName = eventName;
```

Comment:
nit: if it doesn't get changed internally, IMO using `public/private readonly` would be more legible for some of these (eventName, type, and engineList). For example
```ts
export class Lumber {
  constructor(
    public readonly eventName: T,
    public readonly type: LumberType,
    private readonly engineList: ILumberjackEngine[]) {
  }
}
```